### PR TITLE
MGMT-17230: Use IDMS when possible

### PR DESCRIPTION
**Pull request by jhernand**
Currently when we run the 'oc' command in disconnected environments whe generate a `ImageContentSourcePolicy` object inside a YAML file and reference it with the `--icsp-file` command line flag. That is deprecated since OpenShift 4.14, and replaced by the `ImageDigestMirrorSet` object and the `--idms-file` command line flag.

This patch changes the code so that it checks if the `--idms-file` command line flag is available, and if it is it generates a `ImgeDigestMirrorSet` and uses the `--idms-file` flag instead of the deprecated `--icsp-file`.

The previous behaviour is preserved when the `--idms-file` flag isn't available in order to keep supporting older OpenShift versions.

Related: https://issues.redhat.com/browse/MGMT-17230
Related: https://github.com/openshift/enhancements/blob/master/enhancements/api-review/add-new-CRD-ImageDigestMirrorSet-and-ImageTagMirrorSet-to-config.openshift.io.md

## List all the issues related to this PR

- [ ] New Feature 
- [X] Enhancement 
- [ ] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [X] Automation (CI, tools, etc)
- [X] Cloud
- [X] Operator Managed Deployments
- [ ] None

## How was this code tested?

- [ ] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [X] Waiting for CI to do a full test run
- [ ] Manual (Elaborate on how it was tested)
- [x] No tests needed

## Checklist

- [X] Title and description added to both, commit and PR.
- [X] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [X] This change does not require a documentation update (docstring, `docs`, README, etc)
- [X] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- Are the title and description (in both PR and commit) meaningful and clear?
- Is there a bug required (and linked) for this change?
- Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

**Comment by jhernand** [comment]
@CrystalChun please take a look.

File: internal/oc/release_test.go [hunk 1]
```diff
@@ -711,8 +890,113 @@ var _ = Describe("GetReleaseBinaryPath", func() {
...
+		ctrl = gomock.NewController(GinkgoT())
+		mockExecuter = executer.NewMockExecuter(ctrl)
+	})
+
+	AfterEach(func() {
+		ctrl.Finish()
+	})
+
+	DescribeTable(
+		"With correct help text",
```
Comment (paul-maidment) [inline internal/oc/release_test.go:909]: I think we should have a negative test case here too

Especially:
- What if we supply a help command that returns an error (maybe the executable is not installed for example?)
- What happens if the output is not as expected, do we get the correct results?

File: internal/oc/release_test.go [hunk 2]
```diff
@@ -570,45 +572,222 @@ var _ = Describe("getIcspFileFromRegistriesConfig", func() {
...
+			Expect(mirrorsFlag).ShouldNot(BeNil())
+			Expect(mirrorsFlag.file).ShouldNot(BeEmpty())
+			data, err := os.ReadFile(mirrorsFlag.file)
+			Expect(err).ShouldNot(HaveOccurred())
+			Expect(data).Should(MatchYAML(dedent.Dedent(`
+				apiVersion: config.openshift.io/v1
+				kind: ImageDigestMirrorSet
+				metadata:
+				  creationTimestamp: null
+				  name: image-policy
```
Comment (paul-maidment) [inline internal/oc/release_test.go:654]: Am I being picky to suggest that even in the test, we should not call this a policy?

Same comment for all other naming within ImageDigestMirrorSet tests

File: internal/oc/release.go [hunk 3]
```diff
@@ -513,8 +652,62 @@ func getIcspContents(mirrorConfig []mirrorregistries.RegistriesConf) ([]byte, er
...
 }
 
-func removeIcspFile(filename string) {
-	if filename != "" {
-		os.Remove(filename)
-	}
+// getFlagsFromHelp gets the names of the flags of the given command. To do so it removes all the flags from the
+// command, executes it with the '--help' flag and parses the output, assuming that it has the format used by the 'oc'
+// command. It returns an array with the names of the flags in alphabetical order.
+func getFlagsFromHelp(log logrus.FieldLogger, executer executer.Executer, command string) (result []string, err error) {
```
Comment (paul-maidment) [inline internal/oc/release.go:658]: I think I would like to see more tests around extraction of flags from help specifically, just so we can see that edge cases have been covered. 

I know that we potentially cover this in tests later though, so not sure if this is excessive.

This might be a case of testing this function and then non longer having to worry about edge cases in tests later.

**Review by paul-maidment — CHANGES_REQUESTED** [review]
Looks mostly good actually, I just have some questions around tests.

File: internal/oc/release_test.go (hunk 2 above)
Comment (jhernand) [reply internal/oc/release_test.go:654]: Will rename it to "image-mirror-set". Does that look better? Any other suggestion? Note that these (both ICSP and IDMS) are written to temporary files, only for use while running the various "oc" commands, after that they are deleted. So the name isn't really important (unless I miss something).

File: internal/oc/release_test.go (hunk 1 above)
Comment (jhernand) [reply internal/oc/release_test.go:909]: I added a few tests, including one to verify the behavior when the command doesn't exist and another to verify the behavior when the command returns an exit code other than zero. Let me know what you think.

File: internal/oc/release.go (hunk 3 above)
Comment (jhernand) [reply internal/oc/release.go:658]: I added some more cases, please take a look again.

**Comment by jhernand** [comment]
/hold

Holding while I manually verify it.

**Comment by jhernand** [comment]
@paul-maidment I had to rebase this in order to use the `Dedent` function introduced in a a separate pull request (see #6781), so you `/lgtm` is gone. Would you be so kind as to review again?

File: internal/oc/release.go [hunk 4]
```diff
@@ -396,12 +393,7 @@ func (r *release) extractFromRelease(log logrus.FieldLogger, releaseImage, cache
 		return "", err
 	}
 
-	var cmd string
-	if icspFile == "" {
-		cmd = fmt.Sprintf(templateExtract, binary, workdir, insecure, releaseImage)
-	} else {
-		cmd = fmt.Sprintf(templateExtractWithIcsp, binary, workdir, insecure, icspFile, releaseImage)
-	}
+	cmd := fmt.Sprintf(templateExtract, binary, workdir, insecure, mirrorsFlag, releaseImage)
```
Comment (CrystalChun) [inline internal/oc/release.go:396]: Sorry, I've never seen it like this before, but does `Sprintf` expand `mirrorsFlag`? In other words, you don't need to do 
```
cmd := fmt.Sprintf(templateExtract, binary, workdir, insecure, mirrorsFlag.String(), releaseImage)
```

File: internal/oc/release.go [hunk 5]
```diff
@@ -442,21 +444,158 @@ func execute(log logrus.FieldLogger, executer executer.Executer, pullSecret stri
...
 
-	mirrorRegistriesConfig, err := r.mirrorRegistriesBuilder.ExtractLocationMirrorDataFromRegistries()
+	idmsFile, err := os.CreateTemp("", "idms-file")
 	if err != nil {
-		log.WithError(err).Errorf("Failed to get the mirror registries needed for ImageContentSources")
 		return "", err
 	}
+	log.Debugf("Building IDMS file from registries.conf with contents %s", contents)
+	if _, err := idmsFile.Write(contents); err != nil {
+		idmsFile.Close()
```
Comment (CrystalChun) [inline internal/oc/release.go:547]: What do you think about doing this in the function body?
```
defer idmsFile.Close()
```

File: internal/oc/release.go [hunk 6]
```diff
@@ -442,21 +444,158 @@ func execute(log logrus.FieldLogger, executer executer.Executer, pullSecret stri
...
+	if err != nil {
+		return
+	}
+	var (
+		name string
+		file string
+	)
+	if suportsIdmsFileFlag {
+		name = idmsFileFlagName
+		file, err = r.getIdmsFileFromRegistriesConfig(log, mirrorsRegistriesConfig)
```
Comment (CrystalChun) [inline internal/oc/release.go:477]: If `file` comes back empty like [here](https://github.com/openshift/assisted-service/pull/6764/files#diff-3b8a020ef087b81b6b45d56bef6dd8d32fcf9d50dfe3d91fd369755b253cfff8R538) and error is nil, will that affect the formatting of the final [command](https://github.com/openshift/assisted-service/pull/6764/files#diff-3b8a020ef087b81b6b45d56bef6dd8d32fcf9d50dfe3d91fd369755b253cfff8R396)?

**Comment by CrystalChun** [comment]
This is looking good so far! Just a few questions from my end

File: internal/oc/release.go (hunk 4 above)
Comment (jhernand) [reply internal/oc/release.go:396]: No, that explicit conversion isn't needed, the printf-like strings do that conversion automatically. From https://pkg.go.dev/fmt:

> 5. If an operand implements method String() string, that method will be invoked to convert the object to a string, which will then be formatted as required by the verb (if any).

Se an example here: https://go.dev/play/p/3m2nGM1qsIc

File: internal/oc/release.go (hunk 6 above)
Comment (jhernand) [reply internal/oc/release.go:477]: That would result in `--icsp-file=`, without a value. But I believe that can't happen due to how the `getIdmsFileFromRegistrationConfig` is written: it can only return an empty file if the `os.CreateTemp` function returns an empty file and no error. Anyhow I can explicitly check it here if you think it is convenient.

File: internal/oc/release.go (hunk 5 above)
Comment (jhernand) [reply internal/oc/release.go:547]: Yes, makes sense. I tried to not change the existing "getIcspFileFromRegistriesConfig" function, and this is just a copy. I will change it.

File: internal/oc/release.go (hunk 6 above)
Comment (jhernand) [reply internal/oc/release.go:477]: Added a check in this commit: https://github.com/openshift/assisted-service/pull/6764/commits/49952f2106fde101384028e215a4ccaf8ac2236a . Will squash when this is ready to merge.

File: internal/oc/release.go (hunk 5 above)
Comment (jhernand) [reply internal/oc/release.go:551]: Actually after reading this more carefully I think that doing `defer idmsFile.Close()` will not work correctly in all cases: if writing to the file fails it would then try to close the file after removing it, which would fail.

Comment (jhernand) [reply internal/oc/release.go:551]: We could do something like this if we wanted to avoid repeating the `Close` call:

```diff
diff --git a/internal/oc/release.go b/internal/oc/release.go
index a37a6d7b1..b50bb113d 100644
--- a/internal/oc/release.go
+++ b/internal/oc/release.go
@@ -546,13 +546,17 @@ func (r *release) getIdmsFileFromRegistriesConfig(log logrus.FieldLogger,
        if err != nil {
                return "", err
        }
+       defer func() {
+               _ = idmsFile.Close()
+               if err != nil {
+                       _ = os.Remove(idmsFile.Name())
+                       result = ""
+               }
+       }()
        log.Debugf("Building IDMS file from registries.conf with contents %s", contents)
        if _, err := idmsFile.Write(contents); err != nil {
-               idmsFile.Close()
-               os.Remove(idmsFile.Name())
                return "", err
        }
-       idmsFile.Close()
 
        return idmsFile.Name(), nil
 }
```

But that is probably more obscure than the current code.

File: internal/oc/release.go (hunk 4 above)
Comment (CrystalChun) [reply internal/oc/release.go:396]: Very nice, thanks for the explanation!

File: internal/oc/release.go (hunk 6 above)
Comment (CrystalChun) [reply internal/oc/release.go:477]: Hmm what about having a check here instead: https://github.com/openshift/assisted-service/pull/6764/files#diff-3b8a020ef087b81b6b45d56bef6dd8d32fcf9d50dfe3d91fd369755b253cfff8R515
To see if file and name are both set to a string? and return "" if not

E.g.
```
func (i *mirrorsFlagInfo) String() string {
	if i == nil {
		return ""
	}
         if i.name == "" || i.file == "" {
		return ""
	}
	return fmt.Sprintf("--%s=%s", i.name, i.file)
}
```

I'm a bit worried that returning in error will cause the function [here](https://github.com/openshift/assisted-service/pull/6764/files#diff-3b8a020ef087b81b6b45d56bef6dd8d32fcf9d50dfe3d91fd369755b253cfff8R333) to error out, which shouldn't be the case because sometimes there are no mirror registries configured and the command should still run

Comment (jhernand) [reply internal/oc/release.go:477]: No problem, done.

File: internal/oc/release.go [hunk 7]
```diff
@@ -442,21 +444,162 @@ func execute(log logrus.FieldLogger, executer executer.Executer, pullSecret stri
...
+	name string
+
+	// file is the name of a temporary file containing the serialized ImageDigestMirrorSet or
+	// ImageContentSourcePolicy.
+	file string
+}
+
+// String returns the complete text of the flag, something like '--idms-file=/tmp/my-idms.yaml'.
+func (i *mirrorsFlagInfo) String() string {
+	if i == nil || i.name == "" || i.file == "" {
```
Comment (jhernand) [inline internal/oc/release.go:509]: @CrystalChun here ^ is the additional check you suggested.

File: internal/oc/release.go (hunk 5 above)
Comment (CrystalChun) [reply internal/oc/release.go:547]: Oh I see, thanks for looking into it! The original way you had it looks good then

File: internal/oc/release.go (hunk 7 above)
Comment (CrystalChun) [reply internal/oc/release.go:509]: Looks great!

File: internal/oc/release.go [hunk 8]
```diff
@@ -442,21 +444,158 @@ func execute(log logrus.FieldLogger, executer executer.Executer, pullSecret stri
...
 
-// Create a temporary file containing the ImageContentPolicySources
-func (r *release) getIcspFileFromRegistriesConfig(log logrus.FieldLogger) (string, error) {
+// whiteSpaceRE is the regular expression used to split commands into arguments, taking into account that the separator
+// can be one or multiple whilte spaces.
+var whiteSpaceRE = regexp.MustCompile(`\s+`)
 
+// getMirrorsFromRegistriesConfig returns the information needed to generate the command line flag that specifies the
+// mirrors configuration used by the 'oc' command.
+func (r *release) getMirrorsFlagFromRegistriesConfig(log logrus.FieldLogger) (result *mirrorsFlagInfo, err error) {
```
Comment (CrystalChun) [inline internal/oc/release.go:453]:
```suggestion
// getMirrorsFlagFromRegistriesConfig returns the information needed to generate the command line flag that specifies the
// mirrors configuration used by the 'oc' command.
func (r *release) getMirrorsFlagFromRegistriesConfig(log logrus.FieldLogger) (result *mirrorsFlagInfo, err error) {
```

File: internal/oc/release_test.go [hunk 9]
```diff
@@ -570,45 +571,222 @@ var _ = Describe("getIcspFileFromRegistriesConfig", func() {
...
-		mockMirrorRegistriesConfigBuilder.EXPECT().ExtractLocationMirrorDataFromRegistries().Return(regData, nil).Times(1)
-		expected := "apiVersion: operator.openshift.io/v1alpha1\nkind: ImageContentSourcePolicy\nmetadata:\n  creationTimestamp: null\n  name: image-policy\nspec:\n  repositoryDigestMirrors:\n  - mirrors:\n    - host1.example.org:5000/localimages\n    - host1.example.org:5000/openshift\n    source: registry.ci.org\n  - mirrors:\n    - host1.example.org:5000/localimages\n    source: quay.io\n"
-		icspFile, err := oc.getIcspFileFromRegistriesConfig(log)
-		Expect(err).ShouldNot(HaveOccurred())
-		Expect(icspFile).ShouldNot(Equal(""))
-		data, err := os.ReadFile(icspFile)
-		Expect(err).ShouldNot(HaveOccurred())
-		Expect(string(data)).Should(Equal(expected))
-	})
+	Context("With IDSM support", func() {
```
Comment (CrystalChun) [inline internal/oc/release_test.go:574]:
```suggestion
	Context("With IDMS support", func() {
```

File: internal/oc/release_test.go [hunk 10]
```diff
@@ -570,45 +571,222 @@ var _ = Describe("getIcspFileFromRegistriesConfig", func() {
...
 	})
 
-	It("mirror_registries_invalid", func() {
-		mockMirrorRegistriesConfigBuilder.EXPECT().IsMirrorRegistriesConfigured().Return(true).Times(1)
-		mockMirrorRegistriesConfigBuilder.EXPECT().ExtractLocationMirrorDataFromRegistries().Return(nil, fmt.Errorf("extract failed")).Times(1)
-		icspFile, err := oc.getIcspFileFromRegistriesConfig(log)
-		Expect(err).Should(HaveOccurred())
-		Expect(err).Should(MatchError("extract failed"))
-		Expect(icspFile).Should(Equal(""))
+	Context("Without IDSM support, but with deprecated ICSP support", func() {
```
Comment (CrystalChun) [inline internal/oc/release_test.go:684]:
```suggestion
	Context("Without IDMS support, but with deprecated ICSP support", func() {
```

File: internal/oc/release.go (hunk 8 above)
Comment (jhernand) [reply internal/oc/release.go]: Done.

File: internal/oc/release_test.go (hunk 9 above)
Comment (jhernand) [reply internal/oc/release_test.go]: Done.

File: internal/oc/release_test.go (hunk 10 above)
Comment (jhernand) [reply internal/oc/release_test.go]: Done.

**Review by CrystalChun — APPROVED** [review]
/lgtm

Thanks Juan!
